src/System.Text.Primitives/System/Text/FormattingData.cs

```diff
@@ -12,19 +12,309 @@ public struct FormattingData
     {
         private static FormattingData s_invariantUtf16;
         private static FormattingData s_invariantUtf8;
-        private byte[][] _digitsAndSymbols; // this could be flatten into a single array
+        private byte[][] _digitsAndSymbols; // this could be flattened into a single array
+        private List<TrieNode> _parsingTrieList; // The parent nodes of the trie
+        private List<TrieNode> _childTrieNodes; // The child nodes of the trie. These two lists could be flattened into one
+                                                // but I went with this architecture because it is very efficient for the common flat-trie case
         private Encoding _encoding;
 
         public enum Encoding
         {
             Utf16 = 0,
             Utf8 = 1,
         }
+        // used to determine if/how we should construct the parsing trie
+        // eventually, this could include all sorts of information, like encoding rules, etc.
+        public enum ParsingStates
+        {
+            NO_PARSING = 0, // don't construct it
+            PARSING = 1,    // construct it with upfront cost
+            LAZY_PARSING = 3 // construct it but allocate lazy. 3 is chosen so the bitwise check below work
```

Comment:
I think that we may end up needing such a parameter eventually for encodings with special rules. Not positive, but it's definitely something we might need to examine.